src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -370,61 +447,81 @@ public TStructuralType CopyChangedValues(TStructuralType original)
+        {
+            if (original == null)
+            {
+                throw Error.ArgumentNull(nameof(original));
+            }
 
-                    _allProperties[nestedResourceName].SetValue(original, instance);
-                }
-                else
-                {
-                    // Recursively patch the subtree.                    
-                    Contract.Assert(TypedDelta.IsDeltaOfT(((object)deltaNestedResource).GetType()), nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+            if (!_structuredType.IsInstanceOfType(original))
+            {
+                throw Error.Argument(nameof(original), SRResources.DeltaTypeMismatch, _structuredType, original.GetType());
+            }
 
-                    Type newType = deltaNestedResource.StructuredType;
-                    Type originalType = originalNestedResource.GetType();
+            IEnumerable<PropertyAccessor<TStructuralType>> propertiesToCopy = GetUnchangedPropertyNames().Select(s => _allProperties[s]);
+            foreach (PropertyAccessor<TStructuralType> propertyToCopy in propertiesToCopy)
+            {
+                propertyToCopy.Copy(_instance, original);
+            }
 
-                    if (deltaNestedResource.IsComplexType && newType != originalType)
-                    {
-                        originalNestedResource = ReAssignComplexDerivedType(originalNestedResource, newType, originalType, deltaNestedResource.ExpectedClrType);
-                        _structuredType.GetProperty(nestedResourceName).SetValue(original, (object)originalNestedResource);
-                    }
+            CopyUnchangedDynamicValues(original);
+        }
 
-                    deltaNestedResource.CopyChangedValues(originalNestedResource);
-                }
+        /// <summary>
+        /// Overwrites the <paramref name="original"/> entity with the changes tracked by this delta.
+        /// </summary>
+        /// <remarks>The semantics of this operation are equivalent to a HTTP PATCH operation, hence the name.</remarks>
+        /// <param name="original">The entity to be updated.</param>
+        public TStructuralType Patch(TStructuralType original)
+        {
+            if (IsComplexType)
+            {
+                original = ReassignComplexDerivedType(original, _structuredType, original.GetType(), ExpectedClrType) as TStructuralType;
             }
 
-            return original;
+            return CopyChangedValues(original);
+        }
+
+        /// <summary>
+        /// Overwrites the <paramref name="original"/> entity with the values stored in this delta.
+        /// </summary>
+        /// <remarks>The semantics of this operation are equivalent to a HTTP PUT operation, hence the name.</remarks>
+        /// <param name="original">The entity to be updated.</param>
+        public void Put(TStructuralType original)
+        {
+            CopyChangedValues(original);
```

Comment:
We are suppossed to get rid of those changes from the patch method but changing them here will be a breaking change. I think this can be updated in 8.x

Reply:
@habbes @ElizabethOkerio The answer to why we don't need to check if it's a complex type here is because "PUT" overwrites the stored value. In the case of PATCH we're trying to preserve any existing values of properties that are not in the delta